### PR TITLE
refactor(api): remove distinct ColumnExpr API

### PR DESCRIPTION
**Pull request by cpcloud**
Remove `ColumnExpr.distinct()` because it's a confusing API.

This is a breaking change.

Here are a few examples of why this method is easily misused, confusing and error prone:

```
>>> import ibis
>>> t = ibis.table(schema=[("a", "int64"), ("b", "string"), ("c", "string")], name="t")
>>> t.a.distinct().log().distinct()  # what does this even compile to?
>>> t.b.length().distinct() + t.c.distinct().length()  # again, this is very ambiguous if defined at all
```

The `TableExpr.distinct()` method still exists, so it's still possible to
compute distict rows, but it can no longer be done with `t.foo.distinct()`.

Note that this doesn't prevent the following operation from being constructed, but
it will fail to compile due to the expression having multiple leaf tables outside of
a join or correlated subquery:

```
>>> t[["a"]].distinct().a + t["a"]
UnboundTable[r0, name=t]
  a int64
  b string
  c string

Selection[r1]
  table:
    ref: r0
  selections:
    a: int64 = r0.a

Distinct[r2]
  table:
    ref: r1

int64 = Add
  left:
    a: int64 = r2.a
  right:
    a: int64 = r0.a
```

### CODE REVIEW

**Comment by jreback** [comment]
is >  t.a.distinct() still supported? otherwise this is very odd

**Comment by cpcloud** [comment]
> is >  t.a.distinct() still supported? otherwise this is very odd (agree its a bit odd to chain them but couldn't we just detect this an raise at compile time)?

Can you rephrase? Hard to tell what you're asking.

**Comment by jreback** [comment]
is the `.distict()` api still available on a column? this seems like a big change if not

**Comment by cpcloud** [comment]
> is the `.distict()` api still available on a column? this seems like a big change if not

Yes, it still exists.

It's a breaking change and is marked as such.

Detection isn't worth the effort IMO. It requires traversing the tree and encoding a set of ad-hoc rules that prevent certain kinds of operations. We'd be play whack-a-mole ad infinitum prevent all the weird things that can be done.

Moreover there's basically only a single thing that's valid and that's projecting a single `DISTINCT` column. Any operations on `t.foo.distinct()` are ambiguous. For example, does `t.foo.distinct() + 1` mean `(t.foo + 1).distinct()` or something else?

It's also not only about chaining. This is a general problem with having a `distinct` method. How does the operation compose? If the operation is too difficult to make compose in an unambiguous way then IMO it's a candidate for removal.

For example, what does

```
col_expr1.distinct() + col_expr2.distinct()
```

mean? `col_expr1` and `col_expr2` can be any arbitrary column expression.

**Comment by jreback** [comment]
> I'm arguing that there's only a single valid, unambiguous operation, and that it can be done without a distinct method on ColumnExpr.

this is saying 2 things, you are allowing `t.distinct()` e.g. a `TableExpr` operation, ok sure but *not* allowing `t.foo.distinct()` e.g. a `ColumnExpr` operation?

**Comment by cpcloud** [comment]
> > I'm arguing that there's only a single valid, unambiguous operation, and that it can be done without a distinct method on ColumnExpr.
> 
> this is saying 2 things, you are allowing `t.distinct()` e.g. a `TableExpr` operation, ok sure but _not_ allowing `t.foo.distinct()` e.g. a `ColumnExpr` operation?

That is correct. Composing `table_expr.distinct()` is much less error prone.

**Comment by cpcloud** [comment]
@kszucs Thoughts?

**Comment by kszucs** [comment]
@cpcloud needs a rebase

**Comment by kszucs** [comment]
@jreback does it look good to you?

**Comment by jreback** [comment]
> Moreover there's basically only a single thing that's valid and that's projecting a single DISTINCT column.

This is a very valid usecase. I agree this *could* be implented by using `table_expr.distinct()` and then projecting. if that is solution then i guess its ok.

**Comment by cpcloud** [comment]
> > Moreover there's basically only a single thing that's valid and that's projecting a single DISTINCT column.
> 
> This is a very valid usecase. I agree this _could_ be implented by using `table_expr.distinct()` and then projecting. if that is solution then i guess its ok.

Yeah, it is a valid usecase. The question is not about whether computing the distinct values in a column is a valid use case, it's about whether to allow a virtually unlimited number of incorrect expression constructions to make a single use case more convenient.

**Comment by saulpw** [comment]
I think this change is a good one.  xref #1711.
